Realm/ObjectStore/impl/transact_log_handler.cpp

```diff
@@ -27,7 +27,110 @@
 using namespace realm;
 
 namespace {
-class TransactLogHandler {
+// A transaction log handler that just validates that all operations made are
+// onces supported by the object store
+class TransactLogValidator {
+    // Index of currently selected table
+    size_t m_current_table = 0;
+
+    // Tables which were created during the transaction being processed, which
+    // can have columns inserted without a schema version bump
+    std::vector<size_t> m_new_tables;
+
+    REALM_NORETURN
+    REALM_NOINLINE
+    void schema_error()
+    {
+        throw std::runtime_error("Schema mismatch detected: another process has modified the Realm file's schema in an incompatible way");
+    }
+
+    bool schema_error_unless_new_table()
+    {
+        if (std::find(begin(m_new_tables), end(m_new_tables), m_current_table) == end(m_new_tables)) {
+            schema_error();
+        }
+        return true;
+    }
+
+protected:
+    size_t current_table() const noexcept { return m_current_table; }
+
+public:
+    // Schema changes which don't involve a change in the schema version are
+    // allowed
+    bool add_search_index(size_t) { return true; }
+    bool remove_search_index(size_t) { return true; }
+
+    // Creating entirely new tables without a schema version bump is allowed, so
+    // we need to track if new columns are being added to a new table or an
+    // existing one
+    bool insert_group_level_table(size_t table_ndx, size_t, StringData)
+    {
+        m_new_tables.push_back(table_ndx);
```

Comment:
Inserting a table at `table_ndx` suggests that the index of each table that had an index greater than that will have their index incremented. Depending on the order that calls are made, this could result in the indexes in `m_new_tables` becoming out of sync with the index the tables end up at.